packages/backend/src/Tabulators/IRV.ts

```diff
@@ -23,94 +42,190 @@ export function IRV(candidates: string[], votes: ballot[], nWinners = 1, randomT
-                // find top (non-zero) rank, if duplicate top ranks mark as overvote
-                if (ballot[candidate.index] == 0) return;
-                if (ballot[candidate.index] == null) return;
-
-                // candidate has rank
-
-                if (topRemainingRank == 0) {
-                    // set initial top rank
-                    topRemainingRankIndex = candidate.index
-                    topRemainingRank = ballot[candidate.index]
-                    isOverVote = false
-                }
-                else if (ballot[candidate.index] < ballot[topRemainingRankIndex]) {
-                    // set new top rank
-                    topRemainingRankIndex = candidate.index
-                    topRemainingRank = ballot[candidate.index]
-                    isOverVote = false
-                }
-                else if (ballot[candidate.index] === ballot[topRemainingRankIndex]) {
-                    // multiple top ranks, mark as over vote
-                    isOverVote = true
-                }
-            })
-            if (isOverVote) {
-                // ballot is overvote
-                overVoteCount += 1
-            }
-            if (topRemainingRank === 0 || isOverVote) {
-                // ballot is exhausted
-                exhaustedVoteCount += 1
-                exhaustedVotes.push(...activeVotes.splice(i, 1))
+
+    let weightedVotes: weightedVote[] = activeVotes.map(vote => ({ weight: Fraction(1), vote: vote, overvote: false }))
+
+    // Initialize candidate vote pools to empty arrays
+    let candidateVotes: weightedVote[][] = Array(summaryData.candidates.length)
+    for (var i = 0; i < candidateVotes.length; i++) {
+        candidateVotes[i] = [];
+    }
+
+    // Initial vote distribution, moves weighted votes into the appropriate candidate pools 
+    distributeVotes(remainingCandidates, candidateVotes, exhaustedVotes, weightedVotes)
+
+    // Set quota based on number of winners and if its proportional
+    let quota = 0
+    if (proportional) {
+        quota = Math.floor(activeVotes.length/(nWinners + 1) + 1)
+    }
+    else {
+        quota = Math.floor(activeVotes.length/2 + 1)
+    }
+
+    while (results.elected.length < nWinners) {
+
+        let roundResults: irvRoundResults = {
+            winners: [],
+            eliminated: [],
+            logs: []
+        }
+
+        let roundVoteCounts = candidateVotes.map((c, i) => ({ index: i, voteCount: addWeightedVotes(c) }))
+
+        let sortedVoteCounts = [...roundVoteCounts].sort((a, b) => {
+            if (a.voteCount !== b.voteCount) {
+                return (b.voteCount - a.voteCount)
             }
-            else {
-                // give vote to top rank candidate
-                roundVoteCounts[topRemainingRankIndex] += 1
+
+            // break tie in favor of candidate with higher score in previous rounds
+            for (let i = results.voteCounts.length - 1; i >= 0; i--) {
```

Comment:
Yes this is the Scottish STV rules for tiebreaking. Seems to apply to IRV as well except for San Francisco and Oakland rules. Looking at it now only thing that might break here is it could look at previous rounds of block IRV, but maybe that's ok.